ACPs/77-reinventing-subnets/README.md

```diff
@@ -46,77 +46,154 @@ For standalone networks like the Avalanche Primary Network, this is done by conn
+    // Restrictions:
+    // - Must not be the Primary Network ID
+    Subnet ids.ID `json:"subnetID"`
+    // Chain where the Subnet manager lives
+    ChainID ids.ID `json:"chainID"`
+    // Address of the Subnet manager
+    Address []byte `json:"address"`
+    // Authorizes this conversion
+    SubnetAuth verify.Verifiable `json:"subnetAuthorization"`
+}
 ```
 
-- `subnetID`, `nodeID`, and `weight` are for the Subnet Validator being added
-- `expiry` is the time after which this message is invalid. After the P-Chain timestamp is past `expiry`, this Avalanche Warp Message can no longer be used to add the `nodeID` to the validator set of `subnetID`. For replay protection, the P-Chain will store used `messageID`s (hash of the entire Avalanche Warp Message). To prevent the P-Chain from having to store an unbounded number of `messageID`s, the `expiry` is required to be no longer than 48 hours in the future of the time the transaction is issued on the P-Chain. After `RegisterSubnetValidatorTx` is accepted, the P-Chain state will store a `minNonce` of `0` corresponding to the Warp `messageID`. The `minNonce` will be used to validate the `SetSubnetValidatorWeightTx` below.
-- `signature` is the raw bytes of the Ed25519 signature over the concatenated bytes of `[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]`. This signature must correspond to the Ed25519 public key that is used for the `nodeID`. This approach prevents NodeIDs from being unwillingly added to Subnets. `balance` is the minimum initial $nAVAX balance that must be attached to the validator serialized as a uint64. The `signature` field will be validated by the P-Chain in Step 2. A Subnet may choose to validate that the `signature` field is well-formed but it is not required.
+Once this transaction is accepted, `AddSubnetValidatorTx` and `RemoveSubnetValidatorTx` are disabled on the Subnet. `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` must be used to manage the Subnet's validator set going forward.
 
-Subnets are responsible for defining the procedure on how to retrieve the above information from prospective validators.
+### Recovering a Subnet
 
-An EVM Subnet may choose to implement this step like so:
+In the event that a Subnet has no validators, a valid BLS multi-signature cannot be produced. This situation can arise immediately after a Subnet is created or if all the Subnet's validators were removed from the validator set via `SetSubnetValidatorWeightTx` with `Weight = 0`.
 
-- Use the number of tokens the user has staked into a smart contract on the Subnet to determine the weight of their validator
-- Require the user to submit an on-chain transaction with their validator information
-- Generate the Warp message
+A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the `Owner` key defined in `CreateSubnetTx`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued.
```

Comment:
Is `SubnetAuth` in `RecoverSubnetTx` and `ConvertSubnetTx` the same as `Owner`?

Reply:
`SubnetAuth` is necessary for any transactions that require signing with an Owner Key. Many of the existing p-chain transactions already implement this, i.e. [CreateChainTx](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/create_chain_tx.go#L50). Here's an [example](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/wallet/chain/p/signer/visitor.go#L96) made with the p chain wallet.

Reply:
Thanks for the context. I commented because it wasn't clear to me how `Owner` maps to the code. Any docs that describe the relationship between `Owner` and `SubnetAuth` would be helpful.

Reply:
definitely need better docs on this. afaict, the [`verifySubnetAuthorization`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/executor/subnet_tx_verification.go#L53) uses [`getSubnetOwner`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/state/diff.go#L279) to get the owner from state. This was set by [`setSubnetOwner`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/executor/standard_tx_executor.go#L107) when `CreateSubnetTx` was executed.